### DISCUSSION TITLE
Progress Incorporating Unique Vehicles Into Webots

### DISCUSSION THREAD

**Discussion by fradog** [cyberbotics/webots#6332] [General]
I am a retired car designer from a major European Manufaturer and hope to further my interest in designing vehicles which improve the efficacy of our transportation systems.

I feel like I have made some good progress toward a tool / workflow for vehicle designers interested in future mobility solutions. After a steep learning curve I have managed to get a simple model of a vehicle designed using Fusion 360 into the Webots environment with a look that is acceptable to share with fellow design colleagues with a reasonably high standard for visualization and aesthetics. I am thrilled that this is a physics based simulation of potentially real world traffic situations.

My first milestone can be seen here:

https://youtu.be/N4ZJ24MGUI8

It's a simulation using the highway_overtake example. Many trials and errors to get there, but it does work. I have altered the PID coefficients somewhat blindly.

I now need to advance my understanding of the controller logic, and being an old guy with PTSD from Fortran IV experiences 45 years ago in university, I am just beginning to decipher the various codes. It would be helpful if there was documentation available for the highway_overtake controller in python that is aimed at those without much coding experience. More documentation for the other automotive related controllers would also be great. Maybe I haven't looked in the right places yet,  but if there are any suggestions, or other comments they'd be greatly appreciated.

Thanks

**Comment by omichel** [comment]
I would suggest to setup a GitHub repository with you simulation models and try to team up with some experienced software developer which you may find on the [Webots Discord channels](https://discordapp.com/invite/nTWbN9m) or here.

**Reply by fradog** [reply]
Great suggestion, I will follow that up. Thanks! 

Until then, with the ambition to simulate vehicles travelling in bicycle Lanes in addition to on roads, I have discovered that Webots doesn't respect that way type from OSM.

I have manually edited the osm file by replacing v='cycleway' with v='residential', and the ways showed up in Webots. But for Sumo to build the routing simulation with bicycles, I have to use the original files where the cycleways are defined as such. Sumo netconvert does a good job of representing the originally written cycleways, not to mention tricky junctions, so I used that to create a net.net.xml file, and then renamed it to sumo.net.xml after creating a new set of .rou.xml and .sumocfg files in SUMO. This worked with the problem that the vehicles got translated a huge amount relaative to the .wbt file. I then found the line in the sumo.net.xml file which was throwing them off:
<location netOffset=...
I was able to replace the values with the values from the original sumo.net.xml file created with netconvert from the webots generated edg and nod files. This put the vehicles back in the correct location. It's a workaround that is confusing but does work.
I have not been able to locate the source of this <location netOffset=  calculation. My hope was to find it and figure out why the discepancy between the two files is happening. Any help in locating that and/or finding a less confusing solution would be greatly appreciated. Maybe the answer is to find a skilled developer to help me, as it seems the Webots sumo exporter is perhaps outdated or wasn't intended for this purpose.

Also, I find it remarkable that you give us as much attention as you do, Thank you!

**Reply by omichel** [reply]
> I have manually edited the osm file by replacing v='cycleway' with v='residential', and the ways showed up in Webots.

I believe that modifying [that script](https://github.com/cyberbotics/webots/blob/master/resources/osm_importer/webots_objects/road.py#L239) and maybe also [this one](https://github.com/cyberbotics/webots/blob/master/resources/osm_importer/webots_objects/speed_limit.py#L74) may fix the problem once for all.

For the `<location netOffset=` problem, I am not sure where we could fix it. Maybe this could be done in the [sumo supervisor code](https://github.com/cyberbotics/webots/tree/master/projects/default/controllers/sumo_supervisor)?
